disktable.go

```diff
@@ -0,0 +1,650 @@
+			}
+			return nil
+		},
+	)
+	if err != nil {
+		return nil, err
+	}
+	return v, nil
+}
+
+// Len() returns the number of entries in the table.
+func (t *Table) Len() uint64 {
+	return t.primary.counter.Load()
+}
+
+// Lookup provides the Index name and the Value that needs to match for the entry
+// to be returned.
+type Lookup struct {
+	// IndexName is the name of the index to do the lookup in.
+	IndexName string
+	// Key is the key in the index to lookup.
+	Key []byte
+}
+
+// Result is the result of a table lookup.
+type Result struct {
+	Value []byte
+	Err   error
+}
+
+type btreeItem struct {
+	primaryKey uint64
+	count      uint16
+}
+
+func (a btreeItem) Less(than btree.Item) bool {
+	b := than.(btreeItem)
+	return a.primaryKey < b.primaryKey
+}
+
+// Fetch retrieves specifc rows that match all index lookups. You must supply at least
+// 1 lookup. You cannot currently specify multiple searches in the same index.
+// If you wish to fetch all rows, use FetchAll().
+func (t *Table) Fetch(ctx context.Context, lookup ...Lookup) (chan Result, error) {
+	indexesSearched := map[string]bool{}
+
+	for _, l := range lookup {
+		if t.indexByName[l.IndexName] == nil {
+			return nil, fmt.Errorf("index %q is not found", l.IndexName)
+		}
+		if indexesSearched[l.IndexName] {
+			return nil, fmt.Errorf("index %q cannot be used in a search more than once", l.IndexName)
+		}
+		indexesSearched[l.IndexName] = true
+	}
+
+	if len(lookup) == 0 {
+		return nil, fmt.Errorf("Fetch() requires lookups, otherwise use FetchAll()")
+	}
+
+	ch := make(chan Result, 1)
+
+	go func() {
+		defer close(ch)
+
+		ctx, cancel := context.WithCancel(ctx)
+		defer cancel()
+
+		wg := sync.WaitGroup{}
+		mu := sync.Mutex{}
+		matches := btree.New(2)
+
+		// fetch all rows for each index that match our lookup constraint.
+		// Record those rows into a 2-3-4 tree. Any entry in the tree that has
+		// len(lookup) as its value is a line we want to return.
+		errCh := make(chan error, 1)
+		for _, l := range lookup {
+			l := l
+
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				for result := range t.indexLookup(l.IndexName, l.Key) {
+					if ctx.Err() != nil {
+						return
+					}
+					if result.err != nil {
+						if result.err == badger.ErrKeyNotFound {
+							continue
+						}
+						cancel()
+						select {
+						case errCh <- result.err:
+						default:
+						}
+					}
+					// Add entry to btree or increase existing entry counter.
+					mu.Lock()
+
+					var counter btreeItem
+					a := matches.Get(btreeItem{primaryKey: result.i})
+					if a == nil {
+						counter = btreeItem{primaryKey: result.i}
+					} else {
+						counter = a.(btreeItem)
+					}
+					counter.count++
+					matches.ReplaceOrInsert(counter)
+
+					mu.Unlock()
+				}
+			}()
+		}
+
+		wg.Wait()
+
+		// Did we have any errors, if so return them.
+		select {
+		case err := <-errCh:
+			ch <- Result{Err: err}
+			return
+		default:
+		}
```

Comment:
does it make sense to add a `case` for `<- ctx.Done()` here, or is it the caller's responsibility to do so? if the latter, probably should go into the docs.

Reply:
Neither really.  

This select is simply to allow us to capture an error if we have one. If we have already received an error from another goroutine, this simply drops that error since we are only going to return one of them.

Since this is a "push" or "drop", select, it doesn't need to check `ctx.Done()`.

Above in the loop we do `ctx.Err()` to check for an error on every processing of an index return.  So that is where we are doing our context cancellation check.

We do `cancel()`  if we receive an error right above the select in order to cause all our other goroutines to stop since if we have a problem in one index, none of our results can then be trusted.  

The caller only has a responsibility to call cancel() if they pass us a context that has a cancel on it.  Our context cancel is derived from the context passed, so it is local to our method and is our responsibility to call, which happens in a defer statement.

However, what I probably should have done to make this easier to digest is lean on "golang.org/x/sync/errgroup".  It basically encapsulates the waitgroup select{} error stuff inside.  I always forget to use it because I was writing that code out by hand long before it was around.  So I moved to that to make it easier to read and the function was too long anyways, so I a divided it up some.